### PR TITLE
[AWS] Update catalog from host periodically

### LINKED ISSUE #1437: [Spot] Use real time price for the spot VMs
From our user:
> Seems that fetching spot prices from cached catalog rather than live data might have costed us more last week (I’m not 100% sure, but a lot of instances were launched in us-east-2). Curious what’s the timeline to utilize live spot prices? Also if the command line tool can display live prices it’d be helpful.

### PR DESCRIPTION
**Pull request by Michaelvll**
Closes #1437, #1461
This PR simplifies #1449 (removes the config). It will update the local catalog cache periodically (every 7 hours) so that the price can be up-to-date.

It should be merged together with https://github.com/skypilot-org/skypilot-catalog/pull/4

Tested:
- [x] run `tests/run_smoke_tests.sh`

### CODE REVIEW

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py [hunk 1]
```diff
@@ -13,23 +13,27 @@
...
     'eu-west-3',
     'eu-north-1',
-    'me-south-1',
+    # 'me-south-1',
     # 'me-central-1', # failed for no credential
-    'af-south-1',
-    'ap-east-1',
-    'ap-southeast-3',
+    # 'af-south-1',
+    # 'af-south-2', # disabled
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/data_fetchers/fetch_aws.py:34]: Q: does this comment apply to L33-34, and L37's comment apply to L35-37?

Patterns we have used:
```
    # 'af-south-1',  # apply to this line
    # 'af-south-2',  # another per-line comment
```
or

```
    # disabled (apply to one block):
    # 'af-south-1', 
    # 'af-south-2',  
    # another block:
    # 'xx-south-1', 
    # 'xx-south-2',  
```

File: sky/clouds/service_catalog/common.py [hunk 2]
```diff
@@ -46,7 +49,8 @@ def get_catalog_path(filename: str) -> str:
     return os.path.join(_CATALOG_DIR, filename)
 
 
-def read_catalog(filename: str) -> pd.DataFrame:
+def read_catalog(filename: str,
+                 update_frequency_hours: Optional[int] = None) -> pd.DataFrame:
     """Reads the catalog from a local CSV file.
 
     If the file does not exist, download the up-to-date catalog that matches
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/common.py:56]: Add some comment on `update_frequency_hours`?

File: sky/clouds/service_catalog/aws_catalog.py [hunk 3]
```diff
@@ -11,8 +11,12 @@
 if typing.TYPE_CHECKING:
     from sky.clouds import cloud
 
-_df = common.read_catalog('aws/vms.csv')
-_image_df = common.read_catalog('aws/images.csv')
+_UPDATE_FREQUENCY_HOURS = 7
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:14]: nit: add a comment, and a note on keeping this the same as skypilot-catalog's update frequency?

nit: (minor preference) consider replacing 'update' with 'pull'.

File: sky/clouds/service_catalog/common.py [hunk 4]
```diff
@@ -55,23 +59,52 @@ def read_catalog(filename: str) -> pd.DataFrame:
...
-        os.makedirs(os.path.dirname(catalog_path), exist_ok=True)
-        with open(catalog_path, 'w') as f:
-            f.write(r.text)
-        logger.info(f'A new {cloud} catalog has been downloaded to '
-                    f'{catalog_path}')
+
+    meta_path = os.path.join(_CATALOG_DIR, '.meta', filename)
+    os.makedirs(os.path.dirname(meta_path), exist_ok=True)
+
+    def _need_update() -> bool:
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/common.py:66]: Consider adding a comment `# Must be called under a FileLock below.` or move L88 above here. (Was a bit surprised.)

File: sky/clouds/service_catalog/common.py [hunk 5]
```diff
@@ -55,23 +59,52 @@ def read_catalog(filename: str) -> pd.DataFrame:
...
+                    with ux_utils.print_exception_no_traceback():
+                        raise e
+            # Save the catalog to a local file.
+            os.makedirs(os.path.dirname(catalog_path), exist_ok=True)
+            with open(catalog_path, 'w') as f:
+                f.write(r.text)
+            with open(meta_path + '.md5', 'w') as f:
+                f.write(hashlib.md5(r.text.encode()).hexdigest())
+            logger.info(f'{cloud} catalog has been updated (every '
+                        f'{update_frequency_hours}h): {catalog_path}')
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/common.py:107]: Print the paren only if update_frequency_hours is not None?

**Review by concretevitamin — COMMENTED** [review]
Awesome @Michaelvll - some minor comments.

**Comment by Michaelvll** [comment]
Just added the AZ ID to AZ name mapping in this PR, so that we don't need to bump the catalog version again for that fix. PTAL.

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py [hunk 6]
```diff
@@ -274,12 +278,23 @@ def get_all_regions_images_df() -> pd.DataFrame:
...
+
+
 if __name__ == '__main__':
     parser = argparse.ArgumentParser()
     parser.add_argument(
         '--all-regions',
         action='store_true',
         help='Fetch all global regions, not just the U.S. ones.')
+    parser.add_argument(
+        '--no-az-mappings',
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/data_fetchers/fetch_aws.py:295]: Why don't we default to always fetch?

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py [hunk 7]
```diff
@@ -274,12 +278,23 @@ def get_all_regions_images_df() -> pd.DataFrame:
...
+        action='store_true',
+        help='Not fetching the mapping of availability zone ID to name.')
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/data_fetchers/fetch_aws.py:297]:
```suggestion
        help='Do not fetch the mapping from availability zone IDs to zone names.')
```
typically it's more common to not use double negative. consider --az-mappings?

**Review by concretevitamin — COMMENTED** [review]
Some comments; thanks @Michaelvll.

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py [hunk 8]
```diff
@@ -108,7 +111,8 @@ def get_spot_pricing_table(region: str) -> pd.DataFrame:
     for response in response_iterator:
         ret = ret + response['SpotPriceHistory']
     df = pd.DataFrame(ret)[['InstanceType', 'AvailabilityZone', 'SpotPrice']]
-    df = df.set_index(['InstanceType', 'AvailabilityZone'])
+    df = df.rename(columns={'AvailabilityZone': 'AvailabilityZoneName'})
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/clouds/service_catalog/data_fetchers/fetch_aws.py:128]: Added. Thanks!

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py (hunk 6 above)
Comment (Michaelvll) [reply sky/clouds/service_catalog/data_fetchers/fetch_aws.py]: We default to always fetch, but we need to disable it for the catalog repo, as we don't want to publish it.

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py (hunk 7 above)
Comment (Michaelvll) [reply sky/clouds/service_catalog/data_fetchers/fetch_aws.py]: Good point! I made it `az_mapping` and default to `True`.

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py [hunk 9]
```diff
@@ -292,9 +313,14 @@ def fetch_availability_zone_mappings() -> pd.DataFrame:
...
     parser.add_argument(
-        '--no-az-mappings',
+        '--az-mappings',
+        dest='az_mappings',
         action='store_true',
-        help='Not fetching the mapping of availability zone ID to name.')
+        help='Fetch the mapping from availability zone IDs to zone names.')
+    parser.add_argument('--no-az-mappings',
+                        dest='az_mappings',
+                        action='store_false')
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/data_fetchers/fetch_aws.py:322]: Q/nit: does argparse not automatically add this? absl does for every bool arg.

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py (hunk 6 above)
Comment (concretevitamin) [reply sky/clouds/service_catalog/data_fetchers/fetch_aws.py]: I see!

**Review by concretevitamin — APPROVED** [review]
LGTM

File: sky/clouds/service_catalog/data_fetchers/fetch_aws.py (hunk 9 above)
Comment (Michaelvll) [reply sky/clouds/service_catalog/data_fetchers/fetch_aws.py:322]: No, unfortunately, `argparse` requires manually adding the negative option. https://stackoverflow.com/questions/15008758/parsing-boolean-values-with-argparse

**Comment by Michaelvll** [comment]
Labeling this PR to be `do not merge`, as this should go in after https://github.com/skypilot-org/skypilot-catalog/pull/4

**Comment by concretevitamin** [comment]
We may need to update https://skypilot.readthedocs.io/en/latest/reference/faq.html#advanced-how-to-make-skypilot-use-all-global-regions. Maybe use a pointer so we don't forget in the future? e.g.,
```
version=$(cat ~/.sky/catalogs/latest)  # need to handle ~/.sky/catalogs did not exist
mkdir -p ~/.sky/catalogs/${version} 
cd ~/.sky/catalogs/${version}
```

File: docs/source/reference/faq.rst [hunk 10]
```diff
@@ -83,13 +83,18 @@ By default, SkyPilot only supports the US regions on different clouds for conven
...
 .. code-block:: bash
 
-  mkdir -p ~/.sky/catalogs/v4
-  cd ~/.sky/catalogs/v4
+  version=$(python -c 'import sky; print(sky.clouds.service_catalog.constants.CATALOG_SCHEMA_VERSION)')
+  mkdir -p ~/.sky/catalogs/${version}
+  cd ~/.sky/catalogs/${version}
   # Fetch all regions for AWS
-  python -m sky.clouds.service_catalog.data_fetchers.fetch_aws --all-regions
+  # Note: By default, SkyPilot already uses all global regions for AWS, and will periodically update the price
```
Comment (concretevitamin) [inline docs/source/reference/faq.rst:90]: Hmm, this seems to conflict with L82?

Comment (Michaelvll) [reply docs/source/reference/faq.rst]: modified it a bit. How does that look?

File: docs/source/reference/faq.rst [hunk 11]
```diff
@@ -79,17 +79,22 @@ To avoid rerunning the ``setup`` commands, pass the ``--no-setup`` flag to ``sky
 (Advanced) How to make SkyPilot use all global regions?
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-By default, SkyPilot only supports the US regions on different clouds for convenience. If you want to utilize all global regions, please run the following command:
+By default, SkyPilot only supports the US regions on different clouds (except for AWS) for convenience. If you want to utilize all global regions, please run the following command:
```
Comment (concretevitamin) [inline docs/source/reference/faq.rst:82]:
```suggestion
By default, SkyPilot supports most global regions on AWS and only supports the US regions on GCP and Azure. If you want to utilize all global regions, please run the following command:
```

File: docs/source/reference/faq.rst [hunk 12]
```diff
@@ -79,17 +79,22 @@ To avoid rerunning the ``setup`` commands, pass the ``--no-setup`` flag to ``sky
...
-  mkdir -p ~/.sky/catalogs/v4
-  cd ~/.sky/catalogs/v4
+  version=$(python -c 'import sky; print(sky.clouds.service_catalog.constants.CATALOG_SCHEMA_VERSION)')
+  mkdir -p ~/.sky/catalogs/${version}
+  cd ~/.sky/catalogs/${version}
   # Fetch all regions for AWS
-  python -m sky.clouds.service_catalog.data_fetchers.fetch_aws --all-regions
+  # Note: SkyPilot already uses all global regions for AWS, and will periodically update the price
+  # manually running this command will disable the automatic update.
+  # python -m sky.clouds.service_catalog.data_fetchers.fetch_aws --all-regions
```
Comment (concretevitamin) [inline docs/source/reference/faq.rst:92]: Do we want to leave this out? Or, say some reason why users may want to run this?

File: docs/source/reference/faq.rst [hunk 13]
```diff
@@ -79,17 +79,22 @@ To avoid rerunning the ``setup`` commands, pass the ``--no-setup`` flag to ``sky
...
 
-  mkdir -p ~/.sky/catalogs/v4
-  cd ~/.sky/catalogs/v4
+  version=$(python -c 'import sky; print(sky.clouds.service_catalog.constants.CATALOG_SCHEMA_VERSION)')
+  mkdir -p ~/.sky/catalogs/${version}
+  cd ~/.sky/catalogs/${version}
   # Fetch all regions for AWS
-  python -m sky.clouds.service_catalog.data_fetchers.fetch_aws --all-regions
+  # Note: SkyPilot already uses all global regions for AWS, and will periodically update the price
+  # manually running this command will disable the automatic update.
```
Comment (concretevitamin) [inline docs/source/reference/faq.rst:91]:
```suggestion
  # Manually running this command will disable the automatic update.
```

**Review by concretevitamin — APPROVED** [review]

File: docs/source/reference/faq.rst (hunk 12 above)
Comment (Michaelvll) [reply docs/source/reference/faq.rst:92]: Good point! Removed it for now.

**Comment by Michaelvll** [comment]
Tested:
- [x] `tests/run_smoke_tests.sh` (except `test_spot` due to the autostop test for controller and `test_azure_start_stop`)

**Pull request merged**